client.go

```diff
@@ -255,47 +289,73 @@ func (c *Client) do(method, urlpath string, doOptions doOptions) (*http.Response
 	}
 
 	httpClient := c.HTTPClient
-	u := c.getAPIPath(urlpath, query, doOptions.unversioned)
-
-	req, err := http.NewRequest(method, u, params)
-	if err != nil {
-		return nil, err
-	}
-	req.Header.Set("User-Agent", c.userAgent)
-	if doOptions.data != nil {
-		req.Header.Set("Content-Type", "application/json")
-	} else if method == "POST" {
-		req.Header.Set("Content-Type", "plain/text")
-	}
-	if c.username != "" && c.secret != "" {
-		req.SetBasicAuth(c.username, c.secret)
-	}
-
-	for k, v := range doOptions.headers {
-		req.Header.Set(k, v)
-	}
+	endpoint := c.getAPIPath(urlpath, query, doOptions.unversioned)
 
+	// The doOptions Context is shared for every attempted request in the do.
 	ctx := doOptions.context
 	if ctx == nil {
 		ctx = context.Background()
 	}
 
-	resp, err := httpClient.Do(req.WithContext(ctx))
-	if err != nil {
-		// If it is a custom error, return it. It probably knows more than us
-		if serror.IsStorageOSError(err) {
+	for _, address := range c.addresses {
```

Comment:
If this `address` fails, we should move it to the back of the list so the next request doesn't hit it (it will just hang before picking the next one, just like this request).

We should still maintain the "only try each host once" semantic though and return an error if none of them work. There will be some concurrency to take care of!